aquadoggo/src/graphql/client/root.rs

```diff
@@ -0,0 +1,131 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+use async_graphql::{Context, Error, Object, Result, SimpleObject};
+use p2panda_rs::document::DocumentId;
+use p2panda_rs::identity::Author;
+use p2panda_rs::storage_provider::traits::StorageProvider;
+
+use crate::db::provider::SqlStorage;
+use crate::db::Pool;
+
+use super::{EntryArgsRequest, EntryArgsResponse};
+
+#[derive(Default, Debug, Copy, Clone)]
+/// The GraphQL root for the client api that p2panda clients can use to connect to a node.
+pub struct ClientRoot;
+
+#[Object]
+impl ClientRoot {
+    /// Return required arguments for publishing the next entry.
+    async fn next_entry_args(
+        &self,
+        ctx: &Context<'_>,
+        #[graphql(
+            name = "publicKey",
+            desc = "Public key that will publish using the returned entry arguments"
+        )]
+        public_key_param: String,
+        #[graphql(
+            name = "documentId",
+            desc = "Document id to which the entry's operation will apply"
+        )]
+        document_id_param: Option<String>,
+    ) -> Result<EntryArgsResponse> {
+        // Parse and validate parameters
+        let document_id = match document_id_param {
+            Some(val) => Some(val.parse::<DocumentId>()?),
+            None => None,
+        };
+        let args = EntryArgsRequest {
+            author: Author::new(&public_key_param)?,
+            document: document_id,
+        };
+
+        // Prepare database connection
+        let pool = ctx.data::<Pool>().map_err(|err| Error::from(err))?;
+        let provider = SqlStorage {
+            pool: pool.to_owned(),
+        };
+
+        provider
+            .get_entry_args(&args)
+            .await
+            .map_err(|err| Error::from(err))
```

Comment:
In order to access the error here I had to require them to be Send + Sync via the storage provider trait definitions.

https://github.com/p2panda/p2panda/pull/324